crates/evm-nonces-recovery/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+            .map_err(|_err| "Failed pre-upgrade state decoding")?;
+        for (account_id, prev_account) in prev_accounts {
+            let account = <Account<R>>::try_get(account_id)
+                .map_err(|_: ()| "There should be no lost accounts")?;
+            ensure!(
+                account.data == prev_account.data,
+                "Account's data should remain unchanged",
+            );
+            ensure!(
+                account.nonce >= prev_account.nonce,
+                "Account's nonce shouldn't decrease",
+            );
+        }
+
+        let accounts_to_recover = Self::accounts_managed_by_evm()
+            .filter_map(|(account_id, _weight)| account_id)
+            .filter(|account_id| {
+                <Account<R>>::try_get(account_id)
+                    .map(|account| account.nonce.is_zero())
+                    .unwrap_or(true)
+            })
+            .count();
+        ensure!(
+            accounts_to_recover == 0,
+            "There should be no accounts left for recovery",
+        );
+        Ok(())
+    }
+}
+
+impl<R, Precompiles> MigrationBrokenNoncesRecovery<R, Precompiles>
+where
+    R: pallet_evm::Config,
+    R: pallet_evm_system::Config<AccountId = H160>,
+    Precompiles: TypedGet,
+    Precompiles::Type: IntoIterator<Item = H160>,
+{
+    /// Checks the contract account state and recovers it if necessary.
+    /// - If the state is missing, recreates it.
+    /// - If the state has nonce = 0, writes the smallest possible valid nonce.
+    ///
+    /// Precompiled contracts in Ethereum usually have nonce = 0 as well. However, since precompiled
+    /// contracts are typically implemented by hooking calls to specific addresses and adding dummy
+    /// state (to ensure they are callable like regular contracts), there's no need for a non-zero
+    /// nonce unless they explicitly perform state-changing operations like `CREATE`. Therefore,
+    /// precompiled contracts MUST NOT be passed here.
+    fn recover(id: &H160, account: &mut Option<AccountInfoOf<R>>) -> Weight {
+        let Some(account) = account.as_mut() else {
+            info!("Account {id} lacks its state");
+            let (nonce, weight) = Self::min_nonce(id);
+            *account = Some(AccountInfo {
+                nonce,
+                data: Default::default(),
+            });
+            return weight;
+        };
+        if !account.nonce.is_zero() {
+            return Default::default();
+        }
+        info!("Account {id} has zero nonce");
+        let (nonce, weight) = Self::min_nonce(id);
+        account.nonce = nonce;
+        weight
+    }
+
+    /// Computes the minimum possible nonce for a given account.
+    fn min_nonce(id: &H160) -> (<R as pallet_evm_system::Config>::Index, Weight) {
+        let nonce = (1..)
+            .find(|&nonce| {
+                let contract_id = contract_address(id, nonce);
+                let is_known_to_evm = AccountCodes::<R>::contains_key(contract_id);
```

Comment:
Note this won't give us 100% accurate results because of self-destructs. If a contract was created by then self-destructed, given the very bug we're fixing here, there will be no trace of it ever existing.

I suggest documenting it in the code as well.

---

That said, maybe we should apply some heuristics here, and only commit to a given nonce once we see a consecutive gap in absent contract addresses?

Reply:
> If a contract was created by then self-destructed, given the very bug we're fixing here, there will be no trace of it ever existing.

Added a comment to `fn restore`:

```rs
/// Bug #1402 made it possible for "Self-destruct" to delete not only the contract code,
/// but also its state with the nonces used. We have no way to restore their nonces here,
/// but in the future, contracts won't be created at such addresses anyway.
```

> That said, maybe we should apply some heuristics here, and only commit to a given nonce once we see a consecutive gap in absent contract addresses?

I haven't made any efforts in this direction yet.

Reply:
You think it's necessary? If not we can wrap it up.

Reply:
While it is possible in theory, we can only care about the list of cases we have at hand, and inspect all them for any self-destructs.